src/data/finset/sigma.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2017 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Yaël Dillies, Bhavik Mehta
+-/
+import data.finset.card
+
+/-!
+# Finite sets in a sigma type
+
+This file defines a few `finset` constructions on `Σ i, α i`.
+
+## Main declarations
+
+* `finset.sigma`: Given a finset `s` in `ι` and finsets `t i` in each `α i`, `s.sigma t` is the
+  finset of the dependent sum `Σ i, α i`
+* `finset.sigma_lift`: Lifts maps `α i → β i → finset (γ i)` to a map
+  `Σ i, α i → Σ i, β i → finset (Σ i, γ i)`.
+-/
+
+open function multiset
+
+variables {ι : Type*} {α β γ : ι → Type*}
+
+namespace finset
+section sigma
+variables (s s₁ s₂ : finset ι) (t t₁ t₂ : Π i, finset (α i))
+
+/-- `s.sigma t` is the finset of dependent pairs `⟨i, a⟩` such that `i ∈ s` and `a ∈ t i`. -/
+protected def sigma : finset (Σ i, α i) := ⟨_, nodup_sigma s.2 (λ i, (t i).2)⟩
+
+variables {s s₁ s₂ t t₁ t₂}
+
+@[simp] lemma mem_sigma {p : sigma α} : p ∈ s.sigma t ↔ p.1 ∈ s ∧ p.2 ∈ t (p.1) := mem_sigma
+
+@[simp] lemma sigma_nonempty : (s.sigma t).nonempty ↔ ∃ x ∈ s, (t x).nonempty :=
+by simp [finset.nonempty]
+
+@[simp] lemma sigma_eq_empty : s.sigma t = ∅ ↔ ∀ x ∈ s, t x = ∅ :=
+by simp only [← not_nonempty_iff_eq_empty, sigma_nonempty, not_exists]
+
+@[mono] lemma sigma_mono (hs : s₁ ⊆ s₂) (ht : ∀ a, t₁ a ⊆ t₂ a) : s₁.sigma t₁ ⊆ s₂.sigma t₂ :=
+λ ⟨i, a⟩ h, let ⟨hi, ha⟩ := mem_sigma.1 h in mem_sigma.2 ⟨hs hi, ht i ha⟩
+
+lemma sigma_eq_bUnion [decidable_eq (Σ a, α a)] (s : finset ι) (t : Π i, finset (α i)) :
+  s.sigma t = s.bUnion (λ i, (t i).map $ embedding.sigma_mk i) :=
+by { ext ⟨x, y⟩, simp [and.left_comm] }
+
+end sigma
+
+section sigma_lift
+variables [decidable_eq ι]
+
+/-- Lifts maps `α i → β i → finset (γ i)` to a map `Σ i, α i → Σ i, β i → finset (Σ i, γ i)`. -/
+def sigma_lift (f : Π ⦃i⦄, α i → β i → finset (γ i)) (a : sigma α) (b : sigma β) :
```

Comment:
Does it make sense to state this functorially instead?
```suggestion
def sigma_lift {m : Type* → Type*} [alternative m] (f : Π ⦃i⦄, α i → β i → m (γ i)) (a : sigma α) (b : sigma β) :
  m (sigma γ)
```
(Untested)

Sure, we lose some universe polymorphism, but we gain the same operation generalized to all monads rate than just finset.

Reply:
We don't have `finset.alternative`, but I can try writing it.

Reply:
The `alternative` generalisation is nice but I think the universe polymorphism is pretty important, especially since we're dealing with finsets in arbitrary types. Perhaps we could have both?

Reply:
As noted on Zulip, passing through the `finset.alternative` that I've written in #10980 means it's noncomputable. A change to the functor library to not fully bundle the types is necessary first.
So yeah I think we should have this for now and leave the generalization as a TODO.